File: solutions/40.combination-sum-ii.js
/*
 * @lc app=leetcode id=40 lang=javascript
 *
 * [40] Combination Sum II
 */

// @lc code=start
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function (candidates, target, start = 0) {
  if (target < 0) return [];
  if (target === 0) return [[]];
  if (start === 0) candidates.sort((a, b) => a - b);

  const res = [];
  for (let i = start; i < candidates.length; ++i) {
    if (i > start && candidates[i] === candidates[i - 1]) continue; //易错ignore i>start
    const cur = candidates[i];
    combinationSum2(candidates, target - cur, i + 1).forEach((subArr) => {
      res.push([cur, ...subArr]);
    });
  }
  return res;
};
// @lc code=end

console.log(combinationSum2([1, 1, 6, 7], 8));

File: solutions/40.combination-sum-ii.rb
#
# @lc app=leetcode id=40 lang=ruby
#
# [40] Combination Sum II
#

# @lc code=start
# @param {Integer[]} candidates
# @param {Integer} target
# @return {Integer[][]}
def combination_sum2(candidates, target)
  candidates.sort!
  ans = []
  tmp = []
  dfs = ->(start) do
    sum = tmp.sum
    if sum >= target
      ans << tmp.dup if sum == target
      return
    end

    (start...candidates.size).each do |i|
       next if i > start && candidates[i] == candidates[i-1]
       tmp << candidates[i]
       dfs.call(i+1)
       tmp.pop
    end
  end
  dfs.call(0)
  ans
end
# @lc code=end

